src/multiversioning.zig

```diff
@@ -0,0 +1,836 @@
+            &self.timeout_statx,
+        );
+    }
+
+    fn on_statx(self: *Multiversion, _: *IO.Completion, result: anyerror!void) void {
+        _ = result catch |e| {
+            self.timeout_statx_previous = .err;
+            self.start_timeout();
+
+            return self.handle_error(e);
+        };
+
+        // Zero the atime, so we can compare the rest of the struct directly.
+        self.timeout_statx.atime = std.mem.zeroes(os.linux.statx_timestamp);
+
+        if (self.timeout_statx_previous == .err or
+            (self.timeout_statx_previous == .previous and !stdx.equal_bytes(
+            os.linux.Statx,
+            &self.timeout_statx_previous.previous,
+            &self.timeout_statx,
+        ))) {
+            log_multiversioning.info("binary {s} - change detected.", .{self.exe_path});
+
+            self.stage = .init;
+            self.read_from_binary(on_read_from_binary_statx);
+        } else {
+            self.stage = .init;
+            self.start_timeout();
+        }
+
+        self.timeout_statx_previous = .{ .previous = self.timeout_statx };
+    }
+
+    fn on_read_from_binary_statx(self: *Multiversion, result: anyerror!void) void {
+        self.start_timeout();
+
+        _ = result catch return;
+    }
+
+    pub fn read_from_binary(self: *Multiversion, callback: ?Callback) void {
+        self.callback = callback;
+        self.open_source_binary_file();
+    }
+
+    fn open_source_binary_file(self: *Multiversion) void {
+        assert(self.stage == .init or self.stage == .ready or self.stage == .err);
+        self.stage = .source_open;
+
+        self.io.openat(
+            *Multiversion,
+            self,
+            on_open_source_binary_file,
+            &self.completion,
+            IO.INVALID_FILE,
+            self.exe_path,
+            os.O.RDONLY,
+            0,
+        );
+    }
+
+    fn on_open_source_binary_file(
+        self: *Multiversion,
+        _: *IO.Completion,
+        result: IO.OpenatError!os.fd_t,
+    ) void {
+        assert(self.stage == .source_open);
+        const fd = result catch |e| return self.handle_error(e);
+
+        self.stage = .source_read;
+        self.source_fd = fd;
+        self.io.read(
+            *Multiversion,
+            self,
+            on_read_source_file,
+            &self.completion,
+            self.source_fd.?,
+            self.source_buffer,
+            0,
+        );
+    }
+
+    fn on_read_source_file(
+        self: *Multiversion,
+        _: *IO.Completion,
+        result: IO.ReadError!usize,
+    ) void {
+        assert(self.stage == .source_read);
+
+        os.close(self.source_fd.?);
+        self.source_fd = null;
+
+        const bytes_read = result catch |e| return self.handle_error(e);
+        const source_buffer = self.source_buffer[0..bytes_read];
+
+        self.stage = .source_swap;
+        self.swap_source_and_target(source_buffer) catch |e| return self.handle_error(e);
```

Comment:
Don't quite follow?

Reply:
When the function started, we asserted that `stage == .source_read`.
Then we transitioned it to `target_update` and called `self.target_update()`.
After `self.target_update`, can we assert that the stage is either `ready` or `err`?

Reply:
Ah, gotcha. Even better, we can `assert(self.stage == .ready);` since `self.target_update()` returns an error (which would become `.err` in the catch handler.)